### PR TITLE
JAMES-2586 PostgresDelegationStore

### PR DESCRIPTION
**Pull request by hungphan227**

### CODE REVIEW

File: server/container/guice/postgres-common/src/main/java/org/apache/james/modules/data/PostgresDelegationStoreModule.java [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+import com.google.inject.Provides;
+import com.google.inject.Scopes;
+import com.google.inject.Singleton;
+import com.google.inject.multibindings.Multibinder;
+
+public class PostgresDelegationStoreModule extends AbstractModule {
+    @Override
+    public void configure() {
+        bind(DelegationStore.class).to(PostgresDelegationStore.class);
+        bind(PostgresDelegationStore.UserExistencePredicate.class).to(PostgresDelegationStore.UserExistencePredicateImplementation.class);
```
Comment (chibenwa) [inline server/container/guice/postgres-common/src/main/java/org/apache/james/modules/data/PostgresDelegationStoreModule.java:44]: The user existance predicate depends on the user repository implementation, thus should be in `PostgresUsersRepositoryModule`. (We might need the PGDelegationStore with LDAP...)

File: server/data/data-postgres/src/main/java/org/apache/james/user/postgres/PostgresUsersDAO.java [hunk 2]
```diff
@@ -141,4 +150,93 @@ public void addUser(Username username, String password) {
...
+    private Mono<Void> addUserToList(Field<String[]> field, Username baseUser, Username targetUser) {
+        String fullAuthorizedUsersColumnName = TABLE.getName() + "." + field.getName();
+        return postgresExecutor.executeVoid(dslContext ->
+            Mono.from(dslContext.insertInto(TABLE_NAME)
+                .set(USERNAME, baseUser.asString())
+                .set(field, DSL.array(targetUser.asString()))
+                .onConflict(USERNAME)
+                .doUpdate()
+                .set(DSL.field(field.getName()),
+                    (Object) DSL.field("array_append(coalesce(" + fullAuthorizedUsersColumnName + ", array[]::varchar[]), ?)",
```
Comment (chibenwa) [inline server/data/data-postgres/src/main/java/org/apache/james/user/postgres/PostgresUsersDAO.java:176]: string concatenation in a SQL query on a field potentially controlled by a user spotted.

Extra care handling this is definitively needed.

File: server/data/data-postgres/src/main/java/org/apache/james/user/postgres/PostgresUsersDAO.java [hunk 3]
```diff
@@ -141,4 +150,93 @@ public void addUser(Username username, String password) {
...
+
+    private Mono<Void> removeUserFromList(Field<String[]> field, Username baseUser, Username targetUser) {
+        return postgresExecutor.executeVoid(dslContext ->
+            Mono.from(createQueryRemoveUserFromList(dslContext, field, baseUser, targetUser)));
+    }
+
+    private UpdateConditionStep<Record> createQueryRemoveUserFromList(DSLContext dslContext, Field<String[]> field, Username baseUser, Username targetUser) {
+        return dslContext.update(TABLE_NAME)
+            .set(DSL.field(field.getName()),
+                (Object) DSL.field("array_remove(" + field.getName() + ", ?)",
```
Comment (chibenwa) [inline server/data/data-postgres/src/main/java/org/apache/james/user/postgres/PostgresUsersDAO.java:208]: Idem.

File: server/data/data-postgres/src/test/java/org/apache/james/user/postgres/PostgresDelegationStoreTest.java [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+import org.apache.james.backends.postgres.PostgresExtension;
+import org.apache.james.core.Username;
+import org.apache.james.user.api.DelegationStore;
+import org.apache.james.user.api.DelegationStoreContract;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.extension.RegisterExtension;
+
+import reactor.core.publisher.Mono;
+
+public class PostgresDelegationStoreTest implements DelegationStoreContract {
```
Comment (chibenwa) [inline server/data/data-postgres/src/test/java/org/apache/james/user/postgres/PostgresDelegationStoreTest.java:31]: Please add

```
    @Test
    void virtualUsersShouldNotBeListed() {
        testee = new CassandraDelegationStore(cassandraUsersDAO, any -> Mono.just(false));
        addUser(BOB);

        Mono.from(testee().addAuthorizedUser(ALICE).forUser(BOB)).block();

        assertThat(cassandraUsersDAO.listReactive().collectList().block())
            .containsOnly(BOB);
    }
```

File: server/data/data-postgres/src/main/java/org/apache/james/user/postgres/PostgresUsersDAO.java (hunk 2 above)
Comment (hungphan227) [reply server/data/data-postgres/src/main/java/org/apache/james/user/postgres/PostgresUsersDAO.java:176]: String fullAuthorizedUsersColumnName = TABLE.getName() + "." + field.getName();

TABLE is constant. field could be AUTHORIZED_USERS or DELEGATED_USERS which are all constant. Therefore, it is not vulnerable.

File: server/data/data-postgres/src/main/java/org/apache/james/user/postgres/PostgresUsersDAO.java (hunk 3 above)
Comment (hungphan227) [reply server/data/data-postgres/src/main/java/org/apache/james/user/postgres/PostgresUsersDAO.java:208]: Similar to above

File: server/data/data-postgres/src/test/java/org/apache/james/user/postgres/PostgresDelegationStoreTest.java (hunk 4 above)
Comment (hungphan227) [reply server/data/data-postgres/src/test/java/org/apache/james/user/postgres/PostgresDelegationStoreTest.java:34]: added

File: server/data/data-postgres/src/main/java/org/apache/james/user/postgres/PostgresUsersDAO.java (hunk 2 above)
Comment (chibenwa) [reply server/data/data-postgres/src/main/java/org/apache/james/user/postgres/PostgresUsersDAO.java:176]: I did miss read.

Yes of course that's ok.

**Review by chibenwa — APPROVED** [review]

**Review by Arsnael — APPROVED** [review]

**Pull request merged**